src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+	 * NOTE: This object can be put directly in a for-each loop.
+	 *
+	 * @return an Iterator instance for the message history
+	 */
+	@Override
+	public Iterator<IMessage> iterator() {
+
+		buildRange();
+
+		return new MessageHistoryIterator(range);
+	}
+
+	/**
+	 * Requests a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will not block the thread during execution. Use {@link MessageHistoryBuilder#get()} to block the
+	 * thread.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public RequestBuffer.RequestFuture<MessageHistory> request() {
+		return RequestBuffer.request(() -> {
+			return this.get();
+		});
+	}
+
+	/**
+	 * Builds a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will block the thread until all messages are received. Use {@link MessageHistoryBuilder#request()}
+	 * to request this message history without blocking.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public MessageHistory get() {
+
+		// Get cache from the channel, sorted
+		cached = channel.messages.stream().sorted(MessageComparator.REVERSED).collect(Collectors.toList());
+
+		buildRange();
+
+		// Initialize messages
+		messages = count < 0 ?
+				new ArrayList<>(MESSAGE_CHUNK_COUNT) :
+				new ArrayList<>(count);
+
+		// Checks for special case for efficiency- if a range is reverse ordered (start earlier in history than end) but
+		// count unlimited, then it is equivalent to switching endpoints and reversing the collection afterwards
+		if (range.isChronological() && count < 0) {
+
+			range = new MessageHistoryRange(channel,
+					uEnd == null ?
+							new MessageHistoryRange.Endpoint(uEndTime, uIncludeEnd) :
+							new MessageHistoryRange.Endpoint(uEnd, uIncludeEnd),
+					uStart == null ?
+							new MessageHistoryRange.Endpoint(uStartTime, uIncludeStart) :
+							new MessageHistoryRange.Endpoint(uStart, uIncludeStart)
+			);
+
+			fetch();
+
+			Collections.reverse(messages);
+		} else {
+			fetch();
+		}
+
+		return new MessageHistory(messages);
+	}
+
+	private void buildRange() {
+		range = new MessageHistoryRange(channel,
+				uStart == null ?
+						uStartTime == null ?
+								MessageHistoryRange.Endpoint.NOW :
+								new MessageHistoryRange.Endpoint(uStartTime, uIncludeStart) :
+						new MessageHistoryRange.Endpoint(uStart, uIncludeStart),
+				uEnd == null ?
+						uEndTime == null ?
+								MessageHistoryRange.Endpoint.CHANNEL_CREATE :
+								new MessageHistoryRange.Endpoint(uEndTime, uIncludeEnd) :
+						new MessageHistoryRange.Endpoint(uEnd, uIncludeEnd)
+		);
+	}
+
+	private void fetch() {
+
+		// fetch first message from the range
+		IMessage last = range.fetchFirstInRange();
+
+		if (last == null) { // Means out of range start
+			messages = new ArrayList<>();
+			return;
+		}
+
+		if (count < 0 || currentCount++ < count) {
+			messages.add(last); // Add first message
```

Comment:
`last` is a variable that just holds the last fetched message- the starting point for the next batch to fetch. I need to add it because fetched chunks do not include the message you pass in as the parameter.